core/src/main/java/com/datastax/oss/driver/api/core/session/SessionBuilder.java

```diff
@@ -650,6 +651,25 @@ public SelfT withApplicationVersion(@Nullable String applicationVersion) {
     return self;
   }
 
+  /**
+   * The metrics registry object for storing driver metrics.
+   *
+   * <p>The default driver metrics framework is DropWizard and does not need an external metrics
+   * registry object to be provided.
```

Comment:
In general, should the MetricsFactory implementations try to create a Registry object, even if one isn't set on the Context? That is what DropWizardMetricsFactory does now. I can do it for Micrometer as well. But unless we add a compile time dependency on SmallRye, I can't spin up a concrete instance for MircoProfile.

The reason I ask is for consistency. We can add SmallRye as a dependency, and then each of the Factory classes can use a registry object if provided, or use a default implementation if not. Or we could require an appropriate registry object to be set on the context. I'm thinking the former approach may be best as it will capture metrics if an application forgets to set a registry object.

Reply:
> We can add SmallRye as a dependency, and then each of the Factory classes can use a registry object if provided, or use a default implementation if not.

Maybe use reflection for that?

Reply:
Looking at this a bit more, it's going to be difficult to really test Micrometer and MicroProfile metrics if an explicit Registry object is not provided. The Factory classes currently do _not_ implement `getMetrics()`, only Dropwizard does. So if applications setup Micrometer or MicroProfile and do not also specify the metric Registry, they won't have a way to access driver metrics collected.

For Dropwizard, even if an application doe not provide a Registry, the application can still access the metrics from `getMterics()`.

@adutra @olim7t Should I add an accessor method to the MetricsFactory, or some other way to access the metrics when an explicit registry is not specified (for Micrometer and MicroProfile)? Or should we rstick with requiring an external Registry object to be provided in these cases, and only allow Dropwizard to not specify a Registry object?

Reply:
> The Factory classes currently do not implement getMetrics()

Implementing `getMetrics()` may be an interesting addition that stems quite naturally from the general trend of opening up to different metrics implementations. However I don't think that it's required to be done within this PR. It can safely be implemented later as a follow-up ticket. Our current work on Spring Boot does not require that either, so I'd suggest that we keep this PR focused on providing an external registry.

> Should I add an accessor method to the MetricsFactory, or some other way to access the metrics when an explicit registry is not specified (for Micrometer and MicroProfile)? Or should we stick with requiring an external Registry object to be provided in these cases, and only allow Dropwizard to not specify a Registry object?

My suggestions are:

1. For consistency, allow all `MetricsFactory` implementations to use the external registry, if provided; if not, they should either instantiate one on the fly, or error out if the external registry is required.
    a. For microprofile, you _might_ get away by instantiating the SmallRye registry by reflection; but throwing an error is fine too imo.
3. For integration tests, always pass an external registry; this would hopefully make verifications of the registry's contents straightforward, even in the absence of a proper `getMetrics()` implementation.

Reply:
What I have in mind for MicroProfile:

```
      Object possibleMetricRegistry = context.getMetricsRegistry();
      if (possibleMetricRegistry == null) {
        // metrics are enabled, but a metrics registry was not supplied to the context;
        // in this case, attempt to instantiate SmallRye's implementation
        Class<?> smallRyeRegistryClass =
            Reflection.loadClass(
                context.getClassLoader(), "io.smallrye.metrics.MetricsRegistryImpl");
        if (smallRyeRegistryClass != null) {
          try {
            possibleMetricRegistry = smallRyeRegistryClass.newInstance();
          } catch (Exception ignored) {
          }
          if (possibleMetricRegistry == null) {
            throw new IllegalArgumentException(
                "Metrics registry object is NULL. Expected registry object to be of type '"
                    + MetricRegistry.class.getCanonicalName()
                    + "'");
          }
        }
      }
```

The code above could be simplified if we move this instantiation logic to `Reflection`. Actually, the logic is already there in `buildFromConfig` but we would need to extract a `buildFromClassName` method out of it.

Reply:
And Micrometer:

```
      // try to get the metric registry from the context
      Object possibleMetricRegistry = context.getMetricsRegistry();
      if (possibleMetricRegistry == null) {
        // metrics are enabled, but a metrics registry was not supplied to the context;
        // use global registry in this case.
        possibleMetricRegistry = io.micrometer.core.instrument.Metrics.globalRegistry;
      }
```

Reply:
The only problem with this approach is access to the Registry if it's not explicitly set on the context. Metrics will be collected, but the reference to the Registry is private inside the Factory. And since there is no `getRegistry()` method on the Factory, client code wouldn't have any way to get at the Metrics being stored.

It's different for Dropwizard, because we have `getMetrics()` on MetricsFactory. So client code doesn't need access to the Registry to get at the stored metrics.

I only realized this when writing the tests for it as I had implemented almost exactly what you suggest, but I wasn't able to get the test to verify the metrics.

Reply:
I've implemented the suggestion to error out for Micrometer and MicroProfile if no external registry is provided. For Dropwizard, I create a registry to keep the behavior the same. See this [commit](https://github.com/datastax/java-driver/pull/1494/commits/247aa46145c5f7afc88b4ab73948aa9ef32dca18)

Reply:
> The only problem with this approach is access to the Registry if it's not explicitly set on the context. Metrics will be collected, but the reference to the Registry is private inside the Factory. And since there is no `getRegistry()` method on the Factory, client code wouldn't have any way to get at the Metrics being stored.

You are right, sorry for not anticipating that. 

That said, unless I'm mistaken in Micrometer the default registry is global, and hence accessible even without `getMetrics()`. So at least for Micrometer I think it's safe to default to the global registry.

For MicroProfile, I'd suggest that we error out as you did already; and if/when `getMetrics()` is implemented for this framework, we could revisit the logic and create a SmallRye registry on the fly.

Reply:
Good point about Micrometer. I'll change it as you suggest.

Reply:
Sorry for not replying earlier, I agree with what you settled on.

> Implementing `getMetrics()` may be an interesting addition that stems quite naturally from the general trend of opening up to different metrics implementations.

As a side note, this is impossible with the current interface because it directly references DropWizard types. But as Alexandre noted, this interface has become less important now that we have the ability to inject an external registry. Its only added value is that it can construct a metric path given a `Node` and a `NodeMetric`. We could always expose that as a separate utility method, all the factories do it the same way (`buildFullName` in the metric updater classes).